File: sdk/swift/Sources/Idenqa/CaptureView.swift
#if os(iOS)
import SwiftUI
import UIKit
import AVFoundation

/// Embeddable native journey. Core remains authoritative for consent, branches,
/// accepted evidence and completion. Requires NSCameraUsageDescription.
public struct CaptureView: View {
    @StateObject private var model: NativeCaptureModel
    @Environment(\.scenePhase) private var scenePhase
    @AccessibilityFocusState private var headingFocused: Bool

    public init(journey: CaptureJourney, bootstrapToken: String? = nil,
                acquisitionPlans: (any CaptureAcquisitionPlanProvider)? = nil,
                qualityAssessor: any CaptureQualityAssessor = AppleImageQualityAssessor()) {
        _model = StateObject(wrappedValue: NativeCaptureModel(journey: journey, bootstrapToken: bootstrapToken,
            acquisitionPlans: acquisitionPlans, qualityAssessor: qualityAssessor))
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 24) {
                Text(verbatim: model.title).font(.largeTitle.bold())
                    .accessibilityAddTraits(.isHeader).accessibilityFocused($headingFocused)
                    .accessibilityIdentifier("capture.heading")
                if let message = model.message { Text(verbatim: message).accessibilityIdentifier("capture.message") }
                if model.busy { ProgressView("Please wait").frame(maxWidth: .infinity) }
                if !model.busy {
                    content
                }
                if model.snapshot.canCancel {
                    Button("Cancel Verification", role: .cancel) { model.cancel() }
                        .frame(minHeight: 44).accessibilityIdentifier("capture.cancel")
                }
            }
            .padding(24).frame(maxWidth: 560, alignment: .leading).frame(maxWidth: .infinity)
        }
        .background(model.screen == .review ? Color.black : Color(uiColor: .systemBackground))
        .foregroundStyle(model.screen == .review ? Color.white : Color.primary)
        .onChange(of: model.screen) { _ in headingFocused = true }
        .onChange(of: scenePhase) { phase in
            if phase == .background { model.lifecycle(active: false) }
            else if phase == .active { model.lifecycle(active: true) }
        }
        .onDisappear { model.suspend() }
        .sheet(isPresented: $model.cameraVisible, onDismiss: { model.cameraDismissed() }) {
            NativeCameraScreen(front: model.requirement?.camera == .front, title: model.captureTitle, side: model.sideLabel,
                completion: { result in model.receivedPhoto(result) }, requirement: model.requirement,
                qualityAssessor: model.qualityAssessor, sequenceCompletion: { model.receivedSequence($0) })
                .ignoresSafeArea().interactiveDismissDisabled()
        }
    }

    @ViewBuilder private var content: some View {
        switch model.screen {
        case .welcome:
            primary("Get Started") { model.load() }
        case .notice:
            if let state = model.noticeState {
                Text(verbatim: state.notice.copy.summary)
                noticeField("Purpose", state.notice.copy.purpose)
                noticeField("Controller", state.notice.controller)
                noticeField("Recipient", state.notice.recipient)
                noticeField("Consequences", state.notice.copy.consequences)
                noticeField("Jurisdiction", state.jurisdiction)
                noticeField("Processing Regions", state.regions.joined(separator: ", "))
                noticeField("Retention", state.retentionReference)
                primary(state.consentRequired ? "Agree & Continue" : "Acknowledge & Continue") { model.respond(refuse: false) }
                Button("I Do Not Agree", role: .destructive) { model.respond(refuse: true) }.frame(minHeight: 44)
            }
        case .choice:
            if let choice = model.choice {
                ForEach(choice.options) { option in
                    primary(option.label) { model.select(option.id) }
                }
            }
        case .prepare:
            Text(verbatim: model.isSelfie ? "Use even lighting and keep your face uncovered." : "Use your original \(model.documentName). Keep all four corners visible and avoid reflections.")
            primary("Start Camera") { model.openCamera() }
        case .review:
            Text(verbatim: model.sideLabel).font(.headline)
            if let data = model.photo?.bytes, let image = UIImage(data: data) {
                Image(uiImage: image).resizable().scaledToFit().frame(maxHeight: 360)
                    .padding(16).background(.black, in: RoundedRectangle(cornerRadius: 24))
                    .accessibilityLabel("Captured photo for review")
            }
            primary("Use Photo") { model.submit() }
            Button("Retake Photo") { model.retake() }.frame(minHeight: 44)
        case .processing:
            ProgressView("Waiting for Core")
            primary("Check Progress") { model.refresh() }
        case .recovery:
            primary("Try Again") { model.refresh() }
            Button("Open Settings") {
                if let url = URL(string: UIApplication.openSettingsURLString) { UIApplication.shared.open(url) }
            }.frame(minHeight: 44)
        case .finished, .refused, .blocked:
            EmptyView()
        }
    }

    private func primary(_ title: String, action: @escaping () -> Void) -> some View {
        Button(action: action) { Text(verbatim: title).frame(maxWidth: .infinity, minHeight: 44) }
            .buttonStyle(.borderedProminent).accessibilityIdentifier("capture.primary")
    }

    private func noticeField(_ label: String, _ text: String) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(verbatim: label).font(.headline)
            Text(verbatim: text)
        }.accessibilityElement(children: .combine)
    }
}

@MainActor final class NativeCaptureModel: ObservableObject {
    enum Screen { case welcome, notice, choice, prepare, review, processing, recovery, finished, refused, blocked }
    @Published var screen: Screen = .welcome
    @Published var title = "Let’s Verify Your Identity"
    @Published var message: String?
    @Published var busy = false
    @Published var cameraVisible = false
    @Published var snapshot: CaptureJourneySnapshot = .idle
    @Published var noticeState: CaptureNoticeState?
    @Published var choice: CaptureDocumentChoice?
    @Published var photo: CapturedArtifact?
    let journey: CaptureJourney
    private var bootstrapToken: String?
    private var operation: Task<Void, Never>?
    private var generation = 0
    private var started = false
    private var backgrounded = false
    private let acquisitionPlans: (any CaptureAcquisitionPlanProvider)?
    let qualityAssessor: any CaptureQualityAssessor
    private(set) var requirement: CaptureRequirement?
    private(set) var documentName = "identity document"
    private(set) var captureTitle = "Take a Photo"
    var sideLabel: String { isSelfie ? "Selfie" : snapshot.currentTask?.artefact.hasSuffix(".document_back") == true ? "Back of document" : "Front of document" }
    var isSelfie: Bool { snapshot.currentTask?.evidenceType == "idenqa.evidence.selfie_image" }

    init(journey: CaptureJourney, bootstrapToken: String?, acquisitionPlans: (any CaptureAcquisitionPlanProvider)? = nil,
         qualityAssessor: any CaptureQualityAssessor = AppleImageQualityAssessor()) {
        self.journey = journey; self.bootstrapToken = bootstrapToken
        self.acquisitionPlans = acquisitionPlans; self.qualityAssessor = qualityAssessor
    }

    func load() {
        perform {
            if let token = self.bootstrapToken { self.snapshot = try await self.journey.start(bootstrapToken: token); self.bootstrapToken = nil }
            else { self.snapshot = try await self.journey.resume() }
            self.started = true
            try await self.route()
        }
    }

    func refresh() {
        if !started { load(); return }
        perform { self.snapshot = try await self.journey.refresh(); try await self.route() }
    }

    func respond(refuse: Bool) {
        guard let noticeState else { return }
        perform {
            self.noticeState = try await self.journey.respondToNotice(refuse ? .refuse : noticeState.consentRequired ? .consent : .acknowledge)
            try await self.route()
        }
    }

    func select(_ id: String) {
        guard let choice else { return }
        perform {
            self.snapshot = try await self.journey.selectDocument(requirementKey: choice.requirementKey, documentType: id)
            try await self.route()
        }
    }

    func openCamera() {
        perform {
            guard self.requirement != nil else { throw IdenqaError.invalidConfiguration }
            let authority = try await self.journey.notice()
            guard authority.accepted, !authority.refused else { throw IdenqaError.stateConflict }
            guard await AVCaptureDevice.requestAccess(for: .video) else { throw IdenqaError.cameraPermissionDenied }
            try Task.checkCancellation()
            _ = await self.journey.applySensitiveScreenProtection()
            self.cameraVisible = true
        }
    }

    func receivedPhoto(_ result: CapturedArtifact?) {
        cameraVisible = false
        guard let result else { return }
        perform {
            guard let requirement = self.requirement, requirement.challenges.isEmpty else { throw IdenqaError.invalidConfiguration }
            _ = try await AcquisitionCoordinator.validate(result, requirement: requirement, assessor: self.qualityAssessor)
            self.photo = result; self.screen = .review; self.title = "Check Your Photo"
            self.message = "Make sure the image is clear before continuing."
        }
    }

    func cameraDismissed() {} // Keep the sensitive-surface lifetime through review.
    func receivedSequence(_ frames: [AcquiredFrame]) {
        cameraVisible = false
        guard let task = snapshot.currentTask, let requirement else { return }
        perform {
            let authority = try await self.journey.notice()
            guard authority.accepted, !authority.refused else { throw IdenqaError.stateConflict }
            self.snapshot = try await self.journey.submitSequence(taskID: task.id, requirement: requirement, frames: frames)
            await self.journey.removeSensitiveScreenProtection()
            try await self.route()
        }
    }
    func retake() { photo = nil; screen = .prepare; openCamera() }

    func submit() {
        guard let task = snapshot.currentTask, let photo else { return }
        perform {
            let authority = try await self.journey.notice()
            guard authority.accepted, !authority.refused else { throw IdenqaError.stateConflict }
            guard let requirement = self.requirement, requirement.challenges.isEmpty else { throw IdenqaError.invalidConfiguration }
            _ = try await AcquisitionCoordinator.validate(photo, requirement: requirement, assessor: self.qualityAssessor)
            self.snapshot = try await self.journey.submit(taskID: task.id, artifact: photo, method: "idenqa.method.live_camera")
            self.photo = nil
            await self.journey.removeSensitiveScreenProtection()
            self.snapshot = try await self.journey.refresh()
            try await self.route()
        }
    }

    func cancel() {
        suspend()
        perform { _ = try await self.journey.cancel(); self.snapshot = await self.journey.currentSnapshot(); try await self.route() }
    }

    func suspend() {
        generation += 1; operation?.cancel(); operation = nil; busy = false
        cameraVisible = false; photo = nil
        Task { await journey.removeSensitiveScreenProtection() }
    }

    func lifecycle(active: Bool) {
        guard started else { return }
        if active { if backgrounded { backgrounded = false; refresh() } }
        else {
            backgrounded = true
            suspend(); screen = .recovery; title = "Your Progress Is Saved"
            message = "Return to continue. Any unfinished photo will need to be retaken."
            Task { _ = await journey.applicationEnteredBackground() }
        }
    }

    private func route() async throws {
        try Task.checkCancellation()
        message = nil; choice = nil
        switch snapshot.status {
        case .cancelled, .expired, .failed, .completed:
            screen = .finished
            title = snapshot.status == .completed ? "Verification Complete" : snapshot.status == .cancelled ? "Verification Cancelled" : snapshot.status == .expired ? "This Link Has Expired" : "Unable to Continue"
            message = snapshot.status == .completed ? "Return to the organisation that requested this check for the result." : nil
        case .processing, .awaitingExternal, .manualReview:
            screen = .processing; title = "Verifying Your Identity"
        case .captureRequired:
            let notice = try await journey.notice()
            try Task.checkCancellation()
            noticeState = notice
            if notice.refused { screen = .refused; title = "You Chose Not to Continue"; return }
            if !notice.accepted { screen = .notice; title = notice.notice.copy.title; return }
            if let pending = try await journey.documentChoice() { choice = pending; screen = .choice; title = "Choose Your Document"; return }
            guard let task = snapshot.currentTask else { screen = .processing; title = "Finishing Capture"; return }
            // An ordinary photograph must not silently stand in for an active challenge.
            guard task.methodOptions.contains("idenqa.method.live_camera") else {
                screen = .blocked; title = "Additional Capture Support Required"
                message = "This session requires an acquisition adapter that is not available in this native screen."; return
            }
            let label = await journey.documentLabel(for: task) ?? "identity document"
            guard let acquisitionPlans, let sessionID = snapshot.verificationID else {
                screen = .blocked; title = "Capture Setup Required"
                message = "The organisation must supply this session’s acquisition plan before capture can begin."; return
            }
            let plan = try await acquisitionPlans.plan(sessionID: sessionID)
            try Task.checkCancellation()
            requirement = try plan.requirement(sessionID: sessionID, task: task)
            guard let requirement, requirement.challenges.isEmpty ? task.requiredAssurances.isEmpty : requirement.challenges.count >= 2 && requirement.camera == .front else {
                screen = .blocked; title = "Additional Capture Support Required"
                message = "This acquisition plan requires a measured challenge adapter."; return
            }
            documentName = label
            screen = .prepare
            title = isSelfie ? "Take a Clear Selfie" : "Capture the \(task.artefact.hasSuffix(".document_back") ? "Back" : "Front") of Your \(label)"
            captureTitle = title
        default:
            screen = .blocked; title = "Unable to Continue"; message = "Contact the organisation that requested this check."
        }
    }

    private func perform(_ body: @escaping @MainActor () async throws -> Void) {
        guard !busy else { return }
        generation += 1; let current = generation; busy = true; message = nil
        operation = Task {
            defer { if self.generation == current { self.busy = false; self.operation = nil } }
            do { try await body() }
            catch {
                guard self.generation == current, !Task.isCancelled else { return }
                self.screen = .recovery; self.title = "Let’s Try That Again"
                self.message = error as? IdenqaError == .cameraPermissionDenied ? "Allow camera access in Settings to continue." : error as? IdenqaError == .captureQuality ? "The photo is not clear enough. Use even lighting, avoid glare, and retake it." : "We couldn’t finish this step. Check your connection and try again."
            }
        }
    }
}

#endif

File: sdk/kotlin/idenqa/src/main/kotlin/dev/idenqa/sdk/CaptureView.kt
package dev.idenqa.sdk

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.net.Uri
import android.provider.Settings
import android.view.View
import android.widget.*
import kotlinx.coroutines.*
import kotlin.coroutines.resume
import kotlin.coroutines.CoroutineContext

/** Native one-task-at-a-time UI. Forward permission results and host lifecycle. */
class CaptureView(
    private val activity: Activity,
    private val journey: CaptureJourney,
    bootstrapToken: String? = null,
    private val acquisitionPlans: CaptureAcquisitionPlanProvider? = null,
    private val qualityAssessor: CaptureQualityAssessor = AndroidImageQualityAssessor(),
) : ScrollView(activity) {
    private val mainDispatcher = object: CoroutineDispatcher() {
        override fun dispatch(context: CoroutineContext,block: Runnable) { activity.runOnUiThread(block) }
    }
    private val scope = CoroutineScope(SupervisorJob() + mainDispatcher)
    private var work: Job? = null
    private var token = bootstrapToken
    private var started = false
    private var snapshot = CaptureJourneySnapshot.Idle
    private var artifact: CapturedArtifact? = null
    private var camera: CaptureCameraView? = null
    private var busy = false
    private var permissionPending = false
    private var requirement: CaptureRequirement? = null
    private var captureTitle = "Take a Photo"
    private var sideLabel = ""
    private var darkSurface = false
    private var generation = 0
    private var frameWaiter: CancellableContinuation<CapturedArtifact>? = null
    private val column = LinearLayout(activity).apply {
        orientation = LinearLayout.VERTICAL
        setPadding(dp(24),dp(24),dp(24),dp(24))
    }

    init {
        setBackgroundColor(android.graphics.Color.WHITE)
        isFillViewport = true
        addView(column,LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.WRAP_CONTENT))
        heading("Let’s Verify Your Identity")
        button("Get Started") { load() }
    }

    private fun load() = perform {
        snapshot = token?.let { journey.start(it) } ?: journey.resume()
        token=null; started=true; route()
    }
    fun refresh() { if(!started) { load(); return }; perform { snapshot=journey.refresh(); route() } }

    private suspend fun route() {
        currentCoroutineContext().ensureActive()
        setSurface(false)
        column.removeAllViews()
        when(snapshot.status) {
            CaptureJourneyStatus.CAPTURE_REQUIRED -> {
                val notice=journey.notice()
                currentCoroutineContext().ensureActive()
                if(notice.refused) { heading("You Chose Not to Continue"); cancellation(); return }
                if(!notice.accepted) {
                    heading(notice.notice.title); text(notice.notice.summary)
                    field("Purpose",notice.notice.purpose); field("Controller",notice.notice.controller)
                    field("Recipient",notice.notice.recipient); field("Consequences",notice.notice.consequences)
                    field("Jurisdiction",notice.jurisdiction); field("Processing Regions",notice.regions.joinToString(", "))
                    field("Retention",notice.retentionReference)
                    button(if(notice.consentRequired) "Agree & Continue" else "Acknowledge & Continue") {
                        perform { journey.respondToNotice(if(notice.consentRequired) CaptureNoticeAction.CONSENT else CaptureNoticeAction.ACKNOWLEDGE); route() }
                    }
                    button("I Do Not Agree") { perform { journey.respondToNotice(CaptureNoticeAction.REFUSE); route() } }
                } else {
                    val choice=journey.documentChoice()
                    if(choice!=null) {
                        heading("Choose Your Document")
                        choice.options.forEach { option -> button(option.label) {
                            perform { snapshot=journey.selectDocument(choice.requirementKey,option.id); route() }
                        } }
                    } else {
                        val task=snapshot.currentTask
                        if(task==null) { heading("Finishing Capture"); button("Check Progress") { refresh() } }
                        else if("idenqa.method.live_camera" !in task.methodOptions) {
                            heading("Additional Capture Support Required")
                            text("This session requires an acquisition adapter that is not available in this native screen.")
                        } else {
                            val label=journey.documentLabel(task) ?: "identity document"
                            val sessionId = snapshot.verificationId ?: throw IdenqaException.NotStarted
                            if (acquisitionPlans == null) {
                                heading("Capture Setup Required"); text("The organisation must supply this session’s acquisition plan before capture can begin.")
                                cancellation(); return
                            }
                            requirement = acquisitionPlans.plan(sessionId).requirement(sessionId,task)
                            currentCoroutineContext().ensureActive()
                            if (requirement!!.challenges.isEmpty() && task.requiredAssurances.isNotEmpty() ||
                                requirement!!.challenges.isNotEmpty() && (requirement!!.challenges.size < 2 || requirement!!.camera != CaptureCamera.FRONT)) {
                                heading("Additional Capture Support Required"); text("This acquisition plan requires a measured challenge adapter.")
                                cancellation(); return
                            }
                            captureTitle = if(task.evidenceType=="idenqa.evidence.selfie_image") "Take a Clear Selfie" else "Capture the ${if(task.artefact.endsWith(".document_back")) "Back" else "Front"} of Your $label"
                            sideLabel = if(task.evidenceType=="idenqa.evidence.selfie_image") "Selfie" else if(task.artefact.endsWith(".document_back")) "Back of document" else "Front of document"
                            heading(captureTitle)
                            text(if(task.evidenceType=="idenqa.evidence.selfie_image") "Use even lighting and keep your face uncovered." else "Use your original $label. Keep all four corners visible and avoid reflections.")
                            button("Start Camera") { openCamera() }
                        }
                    }
                }
            }
            CaptureJourneyStatus.PROCESSING, CaptureJourneyStatus.AWAITING_EXTERNAL, CaptureJourneyStatus.MANUAL_REVIEW -> {
                heading("Verifying Your Identity"); column.addView(ProgressBar(activity)); button("Check Progress") { refresh() }
            }
            CaptureJourneyStatus.COMPLETED -> { heading("Verification Complete"); text("Return to the organisation that requested this check for the result.") }
            CaptureJourneyStatus.CANCELLED -> heading("Verification Cancelled")
            CaptureJourneyStatus.EXPIRED -> heading("This Link Has Expired")
            else -> { heading("Unable to Continue"); text("Contact the organisation that requested this check.") }
        }
        cancellation()
    }

    private fun openCamera() {
        if(activity.checkSelfPermission(Manifest.permission.CAMERA)!=PackageManager.PERMISSION_GRANTED) {
            permissionPending=true; activity.requestPermissions(arrayOf(Manifest.permission.CAMERA),PERMISSION_REQUEST); return
        }
        perform {
            val policy = requirement ?: throw IdenqaException.InvalidConfiguration
            val notice=journey.notice()
            if(!notice.accepted || notice.refused) throw IdenqaException.StateConflict
            currentCoroutineContext().ensureActive()
            journey.applySensitiveScreenProtection()
            column.removeAllViews(); setSurface(true)
            heading(captureTitle)
            text("Fit the entire required side inside the frame. Avoid glare and keep the camera steady.")
            val challenged=policy.challenges.isNotEmpty()
            val instruction=if(challenged) text("Look straight at the camera",22f) else null
            val ring=if(challenged) CapturePoseRing(activity,policy.challenges.size) else null
            val shutter=Button(activity).apply { text="Capture Photo"; isEnabled=false; minHeight=dp(48) }
            val preview=CaptureCameraView(activity,policy.camera == CaptureCamera.FRONT,
                ready={ if(challenged) startLiveness(policy,instruction!!,ring!!) else shutter.isEnabled=true },
                photo={ value ->
                    if(challenged) { val waiter=frameWaiter; frameWaiter=null; if(waiter?.isActive==true) waiter.resume(value) }
                    else receivedPhoto(value)
                }, failure={ frameWaiter?.cancel(); frameWaiter=null; work?.cancel(); busy=false; showRecovery("The camera is unavailable. Try again.") })
            camera=preview
            val frame = FrameLayout(activity).apply {
                background = android.graphics.drawable.GradientDrawable().apply {
                    setColor(android.graphics.Color.BLACK); cornerRadius=dp(24).toFloat(); setStroke(dp(3),android.graphics.Color.WHITE)
                }
                setPadding(dp(3),dp(3),dp(3),dp(3)); clipToOutline=true
                addView(preview,FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.MATCH_PARENT))
                if(ring!=null) addView(ring,FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.MATCH_PARENT))
            }
            column.addView(frame,LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT,dp(280)))
            text(sideLabel)
            button("Help") {
                android.app.AlertDialog.Builder(activity).setTitle("Capture Help")
                    .setMessage("Use even lighting. Keep all four corners visible, hold the camera steady and move away from reflections. For a selfie, keep your face uncovered and the camera at eye level.")
                    .setPositiveButton("Got It",null).show()
            }
            shutter.setOnClickListener { shutter.isEnabled=false; preview.takePhoto() }
            if(!challenged) column.addView(shutter)
            button("Back") { work?.cancel(); busy=false; stopCamera(); refresh() }; cancellation()
        }
    }

    fun onRequestPermissionsResult(requestCode: Int,grantResults: IntArray) {
        if(requestCode!=PERMISSION_REQUEST || !permissionPending) return
        permissionPending=false
        if(grantResults.firstOrNull()==PackageManager.PERMISSION_GRANTED) openCamera()
        else showRecovery("Allow camera access in Settings to continue.")
    }

    private fun receivedPhoto(value: CapturedArtifact) {
        camera?.close(); camera=null
        perform {
            val policy = requirement ?: throw IdenqaException.InvalidConfiguration
            if (policy.challenges.isNotEmpty()) throw IdenqaException.InvalidConfiguration
            withContext(Dispatchers.Default) { validateCapturePhoto(value,policy,qualityAssessor) }
            currentCoroutineContext().ensureActive()
            showPhoto(value)
        }
    }

    private fun showPhoto(value: CapturedArtifact) {
        artifact=value; column.removeAllViews(); setSurface(true); heading("Check Your Photo"); text(sideLabel)
        val image=try { decodeUprightCapture(value,1280) } catch(_: Exception) { artifact=null; showRecovery("The photo could not be read. Try again."); return }
        column.addView(ImageView(activity).apply { setImageBitmap(image); adjustViewBounds=true; contentDescription="Captured photo for review" },
            LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT,dp(360)))
        text("Make sure the image is clear before continuing.")
        button("Use Photo") { submit() }; button("Retake Photo") { artifact=null; openCamera() }; cancellation()
    }

    private fun submit() {
        val task=snapshot.currentTask ?: return; val photo=artifact ?: return
        perform {
            val notice=journey.notice()
            if(!notice.accepted || notice.refused) throw IdenqaException.StateConflict
            val policy = requirement ?: throw IdenqaException.InvalidConfiguration
            if (policy.challenges.isNotEmpty()) throw IdenqaException.InvalidConfiguration
            withContext(Dispatchers.Default) { validateCapturePhoto(photo,policy,qualityAssessor) }
            snapshot=journey.submit(task.id,photo,"idenqa.method.live_camera")
            artifact=null; journey.removeSensitiveScreenProtection(); snapshot=journey.refresh(); route()
        }
    }

    fun onHostStopped() {
        work?.cancel(); busy=false; permissionPending=false; artifact=null; stopCamera()
        if(started) {
            column.removeAllViews(); heading("Your Progress Is Saved"); text("Return to continue. Any unfinished photo will need to be retaken.")
            button("Continue") { refresh() }; cancellation()
            scope.launch { journey.applicationEnteredBackground() }
        }
    }
    fun onHostResumed() { if(started) refresh() }
    fun close() { onHostStopped(); token=null; scope.cancel() }
    private fun stopCamera() { frameWaiter?.cancel(); frameWaiter=null; camera?.close(); camera=null; scope.launch { journey.removeSensitiveScreenProtection() } }

    private fun startLiveness(policy: CaptureRequirement, instruction: TextView, ring: CapturePoseRing) {
        val task=snapshot.currentTask ?: return
        generation++; val current=generation; busy=true
        work=scope.launch {
            val tracker=MediaPipePoseTracker(activity)
            try {
                val source=RawCaptureSource {
                    withContext(mainDispatcher) {
                        suspendCancellableCoroutine { continuation ->
                            val preview=camera
                            if(preview==null) { continuation.cancel(); return@suspendCancellableCoroutine }
                            frameWaiter=continuation
                            continuation.invokeOnCancellation { activity.runOnUiThread { if(frameWaiter===continuation) frameWaiter=null } }
                            preview.takePhoto()
                        }
                    }
                }
                val coordinator=AcquisitionCoordinator(source,qualityAssessor,ChallengePresenter {
                    withContext(mainDispatcher) { instruction.text="Look straight at the camera" }
                },tracker,progress={ id, progress ->
                    withContext(mainDispatcher) {
                        val index=policy.challenges.indexOfFirst { it.id==id }
                        instruction.text=nativePoseInstruction(policy.challenges[index].prompt,progress.feedback)
                        ring.update(index,progress.fraction)
                    }
                })
                val frames=withContext(Dispatchers.Default) { coordinator.acquire(policy) }
                currentCoroutineContext().ensureActive()
                camera?.close(); camera=null
                instruction.text="Sending your capture"
                val authority=journey.notice()
                if(!authority.accepted || authority.refused) throw IdenqaException.StateConflict
                snapshot=journey.submitSequence(task.id,policy,frames)
                stopCamera(); route()
            } catch(error: CancellationException) { throw error }
            catch(_: Exception) { showRecovery("We couldn’t complete the movement. Check your lighting and connection, then try again.") }
            finally { withContext(NonCancellable+Dispatchers.Default) { tracker.close() }; if(generation==current) busy=false }
        }
    }

    private fun perform(operation: suspend () -> Unit) {
        if(busy) return
        generation++; val current=generation
        busy=true
        // Keep cancellation reachable while transport is in flight.
        column.removeAllViews(); heading("Please Wait"); column.addView(ProgressBar(activity)); cancellation()
        work=scope.launch {
            try { operation() }
            catch(error: CancellationException) { throw error }
            catch(error: Exception) { showRecovery(if(error == IdenqaException.CaptureQuality) "The photo is not clear enough. Use even lighting, avoid glare, and retake it." else "We couldn’t finish this step. Check your connection and try again.") }
            finally { if(generation==current) busy=false }
        }
    }
    private fun cancellation() {
        if(snapshot.canCancel) button("Cancel Verification") {
            work?.cancel(); busy=false; artifact=null; stopCamera()
            perform { journey.cancel(); snapshot=journey.currentSnapshot(); route() }
        }
    }
    private fun showRecovery(message: String) {
        stopCamera(); column.removeAllViews(); setSurface(false); heading("Let’s Try That Again"); text(message)
        button("Try Again") { refresh() }
        button("Open Settings") { activity.startActivity(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS,Uri.parse("package:${activity.packageName}"))) }
        cancellation()
    }
    private fun heading(value: String) { text(value,28f).apply { if(android.os.Build.VERSION.SDK_INT>=28) isAccessibilityHeading=true; requestFocus() } }
    private fun field(label: String,value: String) { text(label,18f); text(value) }
    private fun text(value: String,size: Float=16f): TextView = TextView(activity).also {
        it.setTextColor(if(darkSurface) android.graphics.Color.WHITE else android.graphics.Color.BLACK)
        it.text=value; it.textSize=size; it.setPadding(0,dp(8),0,dp(12)); column.addView(it)
    }
    private fun button(value: String,action: () -> Unit) { column.addView(Button(activity).apply {
        text=value; minHeight=dp(48); isAllCaps=false; setOnClickListener { action() }
    },LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.WRAP_CONTENT)) }
    private fun dp(value: Int): Int = (value*resources.displayMetrics.density).toInt()
    private fun setSurface(dark: Boolean) {
        darkSurface=dark
        setBackgroundColor(if(dark) android.graphics.Color.BLACK else android.graphics.Color.WHITE)
        column.setBackgroundColor(android.graphics.Color.TRANSPARENT)
    }
    companion object { const val PERMISSION_REQUEST = 0x1D01 }
}
